pkg/reconciler/shared/tektonconfig/pipeline/pipeline.go

```diff
@@ -20,108 +20,104 @@ import (
+	ok, err := isTektonPipelineReady(tpCR, err)
+	if err != nil {
+		return nil, err
 	}
+	if !ok {
+		return nil, v1alpha1.RECONCILE_AGAIN_ERR
+	}
+
 	return tpCR, err
 }
 
 func GetPipeline(ctx context.Context, clients op.TektonPipelineInterface, name string) (*v1alpha1.TektonPipeline, error) {
 	return clients.Get(ctx, name, metav1.GetOptions{})
 }
 
-// WaitForTektonPipelineState polls the status of the TektonPipeline called name
-// from client every `interval` until `inState` returns `true` indicating it
-// is done, returns an error or timeout.
-func waitForTektonPipelineState(ctx context.Context, clients op.TektonPipelineInterface, name string,
-	inState func(s *v1alpha1.TektonPipeline, err error) (bool, error)) (*v1alpha1.TektonPipeline, error) {
-	span := logging.GetEmitableSpan(ctx, fmt.Sprintf("WaitForTektonPipelineState/%s/%s", name, "TektonPipelineIsReady"))
-	defer span.End()
-
-	var lastState *v1alpha1.TektonPipeline
-	waitErr := wait.PollImmediate(common.Interval, common.Timeout, func() (bool, error) {
-		lastState, err := clients.Get(ctx, name, metav1.GetOptions{})
-		return inState(lastState, err)
-	})
+func CreatePipeline(ctx context.Context, clients op.TektonPipelineInterface, config *v1alpha1.TektonConfig) (*v1alpha1.TektonPipeline, error) {
+	tpCR := &v1alpha1.TektonPipeline{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: v1alpha1.PipelineResourceName,
+		},
+		Spec: v1alpha1.TektonPipelineSpec{
+			CommonSpec: v1alpha1.CommonSpec{
+				TargetNamespace: config.Spec.TargetNamespace,
+			},
+			Pipeline: config.Spec.Pipeline,
+			Config:   config.Spec.Config,
+		},
+	}
+	_, err := clients.Create(ctx, tpCR, metav1.CreateOptions{})
+	if err != nil {
+		return nil, err
+	}
+	return tpCR, err
+}
+
+func UpdatePipeline(ctx context.Context, tpCR *v1alpha1.TektonPipeline, config *v1alpha1.TektonConfig, clients op.TektonPipelineInterface) (*v1alpha1.TektonPipeline, error) {
+	// if the pipeline spec is changed then update the instance
+	updated := false
+
+	if config.Spec.TargetNamespace != tpCR.Spec.TargetNamespace {
+		tpCR.Spec.TargetNamespace = config.Spec.TargetNamespace
+		updated = true
+	}
+
+	if !reflect.DeepEqual(tpCR.Spec.Pipeline, config.Spec.Pipeline) {
+		tpCR.Spec.Pipeline = config.Spec.Pipeline
+		updated = true
+	}
 
-	if waitErr != nil {
-		return lastState, fmt.Errorf("tektonpipeline %s is not in desired state, got: %+v: %w: For more info Please check TektonPipeline CR status", name, lastState, waitErr)
+	if !reflect.DeepEqual(tpCR.Spec.Config, config.Spec.Config) {
+		tpCR.Spec.Config = config.Spec.Config
+		updated = true
+	}
+
+	if tpCR.ObjectMeta.OwnerReferences == nil {
+		ownerRef := *metav1.NewControllerRef(config, config.GroupVersionKind())
+		tpCR.ObjectMeta.OwnerReferences = []metav1.OwnerReference{ownerRef}
+		updated = true
+	}
+
+	if updated {
+		tpCR, err := clients.Update(ctx, tpCR, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, err
+		}
+		return tpCR, nil
 	}
```

Comment:
```suggestion
	if updated {
		tpCR, err := clients.Update(ctx, tpCR, metav1.UpdateOptions{})
		if err != nil {
			return nil, err
		}
		return nil, v1alpha1.RECONCILE_AGAIN_ERR
	}
```

As a general best practice, if the reconciler (or subfuctions) changes the cluster (API data or trigger a state change on cluster), then we reconcile get out and retry.